src/AppInstallerRepositoryCore/Microsoft/ARPHelper.cpp

```diff
@@ -3,11 +3,57 @@
 #include "pch.h"
 #include "ARPHelper.h"
 #include "winget/PortableARPEntry.h"
+#include <Msi.h>
 
 namespace AppInstaller::Repository::Microsoft
 {
     using namespace AppInstaller::Registry::Portable;
 
+    namespace
+    {
+        // Finds the UpgradeCode corresponding to a given ProductCode
+        std::optional<std::string> GetUpgradeCodeByProductCode(const std::string& productCode)
+        {
+            // The UpgradeCode is not stored in the ARP registry keys, so we use the MSI API to query it.
+            //
+            // The UpgradeCode is apparently also stored in the registry under
+            //   HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\Installer\UpgradeCodes\<UpgradeCode>
+            // but:
+            //   * This key does not seem to be documented
+            //   * The format used to store the GUIDs is harder to parse
+            //   * We cannot look up by ProductCode; we need to take an UpgradeCode and then find the ProductCode
+            // Using the registry could potentially be faster, so we may consider using it if this proves to be too slow.

```

Comment:
This turned out to be quite slow once I actually tested it, so I'll look into other ways of getting the UpgradeCode